### PR TITLE
Deliver balanced transcript-first TUI chat mode with fullscreen first-run onboarding

### LINKED ISSUE #689: [Feature]: add a balanced transcript-first TUI mode for interactive chat
## Surface
Conversation / session runtime

## Problem statement
LoongClaw's interactive chat now has a stronger line-oriented terminal surface after the shared screen-spec work in #518 and the live phase surfaces in #531, but it still does not offer a dedicated full-screen chat shell for longer-running interactive sessions. The current path is good for startup summaries, structured replies, and phase updates, yet it cannot keep the transcript, composer, current execution state, and detailed tool output visible within one stable terminal layout.

That gap matters once interactive chat becomes a daily-use operator surface rather than a one-shot terminal exchange. Users need a transcript-first shell that keeps conversation context anchored, keeps tool activity visible without flooding the screen, and only expands into richer execution detail when approvals, errors, or explicit inspection demand it.

## Proposed solution
Add an opt-in balanced TUI mode for interactive chat, exposed as `loongclaw chat --ui tui`, while keeping the current text path as the default and fallback.

Scope the first iteration narrowly:

- keep the implementation Rust-native in `loongclaw-app`
- preserve the existing text/screen-spec surface as `--ui text`
- add a transcript-first full-screen shell for `--ui tui`
- keep the main screen centered on one active foreground session
- show a quiet execution band for running tools, latest result, background count, and pending approvals
- open richer tool output, shell logs, diffs, and approval/error detail in an inline drawer rather than a permanent operator dashboard
- degrade cleanly back to the text path when stdout/stdin are not suitable for a full-screen TUI or terminal capability checks fail

This should build on the existing live-surface/runtime observation work instead of replacing it: the TUI should consume the same typed chat/runtime state and observer signals, but render them through a dedicated terminal shell.

Related but not duplicate:
- #518 unified the shared structured text/screen-spec surface across onboarding and chat
- #531 added phase-aware live chat surfaces within that existing path
- #217 is broader daemon/operator-runtime productization, whereas this issue is specifically the interactive chat shell

## Non-goals / out of scope
- Do not change the default `loongclaw chat` behavior away from the current text path in the first iteration.
- Do not redesign `loongclaw ask` or other one-shot stdout contracts.
- Do not introduce a multi-session/operator-console dashboard in v1.
- Do not move terminal rendering ownership into `loongclaw-daemon`; CLI flag parsing there is enough.
- Do not require a web UI, browser runtime, or external JS stack.
- Do not replace the current shared screen-spec renderer; keep it as the compatible fallback and non-TUI path.

## Alternatives considered
- Continue extending the current shared text/screen-spec surface only. This keeps scope smaller, but it cannot provide a persistent transcript/composer/execution layout without bending a line-oriented renderer into a second job it was not designed for.
- Build an operator-first multi-pane dashboard immediately. This would improve raw observability, but it overweights background state and pulls the product away from the conversation-first interactive chat experience.
- Route the next UX step into a browser or local web launcher instead of a terminal TUI. That may still be useful for onboarding or operator control planes, but it does not solve the need for a native interactive terminal chat shell.

## Acceptance criteria
- `loongclaw chat --ui tui` launches an interactive full-screen terminal shell while `loongclaw chat` and `loongclaw chat --ui text` preserve the current text behavior.
- The TUI keeps the conversation transcript as the primary persistent surface and includes a composer for interactive input.
- Tool/runtime activity remains visible through a compact execution band without forcing a permanent multi-pane dashboard.
- Approval requests and execution errors auto-surface richer detail, and users can inspect tool output or diffs via an inline drawer.
- Non-interactive or incompatible terminal environments degrade safely to the existing text mode instead of failing unpredictably.
- Workspace tests, all-features tests, and clippy remain green after the change.

## Policy / security / breaking sensitivity
No, this looks routine

## Rollout / rollback notes
Land the first iteration behind the explicit `--ui tui` selector only. Keep `--ui text` as the default and compatibility path. If regressions appear, disable or revert the TUI-specific path without disturbing the existing shared text/screen-spec surfaces.

## Impact
Affected users: operators using `loongclaw chat` for sustained interactive sessions
Frequency: every interactive chat session that outgrows the current line-oriented surface
Consequence if missing: LoongClaw keeps improving chat status visibility incrementally, but still lacks a cohesive transcript-first terminal shell for daily interactive use.

## Submission checks
- [x] I searched existing issues and did not find the same request already tracked.
- [x] I used project-scoped wording and removed secrets or private data from examples.

### PR DESCRIPTION
**Pull request by gh-xj**
## Summary

- Problem:
  LoongClaw did not yet deliver a runtime-backed transcript-first fullscreen chat shell for sustained terminal sessions, and the missing-config first-run path still split users out into separate onboarding surfaces instead of keeping setup, transcript, composer, and execution visibility inside one stable TUI.
- Why it matters:
  The operator experience stayed more black-box than it needed to be. Users could not keep conversation context, tool activity, and first success setup inside one coherent terminal layout, which weakened both observability and perceived control.
- What changed:
  This branch delivers the balanced fullscreen TUI chat shell in `loongclaw-app`, keeps live runtime/tool surfaces inside that shell, adds lazy boot-flow plumbing so the fullscreen shell can start before the chat runtime is initialized, implements a first-run fullscreen onboarding state machine for missing-config TUI launches, and routes both `loongclaw chat --ui tui` and `loongclaw tui` into that integrated path when config is absent. It also preserves the chat handoff after a successful config write and adds PTY coverage for missing-config onboarding entry plus successful onboarding-to-chat transition. Follow-up polish in this branch makes the composer truly multiline with `Shift+Enter`, adds empty-composer transcript navigation via `Up/Down`, `Home/End`, and `PageUp/PageDown`, renders a floating `End latest` affordance together with a persistent `LIVE` / `SCROLLED` transcript state indicator, updates the help overlay so the shortcuts stay discoverable, and hardens the onboarding PTY tests by synchronizing on the first visible fullscreen frame before matching onboarding copy.
- What did not change (scope boundary):
  Default `loongclaw chat` still uses the text path. Terminal rendering ownership still stays in `loongclaw-app` rather than moving into `loongclaw-daemon`. `loongclaw onboard` remains available as the richer standalone onboarding flow. This PR does not add a multi-session dashboard or any web UI.

## Linked Issues

- Closes #689
- Related #690

## Change Type

- [x] Bug fix
- [x] Feature
- [x] Refactor
- [x] Documentation
- [ ] Security hardening
- [ ] CI / workflow / release

## Touched Areas

- [ ] Kernel / policy / approvals
- [ ] Contracts / protocol / spec
- [x] Daemon / CLI / install
- [ ] Providers / routing
- [ ] Tools
- [ ] Browser automation
- [ ] Channels / integrations
- [x] ACP / conversation / session runtime
- [ ] Memory / context assembly
- [x] Config / migration / onboarding
- [x] Docs / contributor workflow
- [ ] CI / release / workflows

## Risk Track

- [x] Track A (routine / low-risk)
- [ ] Track B (higher-risk / policy-impacting)

## Validation

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --workspace --all-targets --all-features -- -D warnings`
- [x] `cargo test --workspace --locked`
- [x] `cargo test --workspace --all-features --locked`
- [x] Relevant architecture / dep-graph / docs checks for touched areas
- [x] Additional scenario, benchmark, or manual checks when behavior changed
- [x] If this changes config/env fallback, limits, or defaults: include before/after behavior and regression coverage for explicit path, fallback path, and boundary values
- [x] If tests mutate process-global env: document how state is restored or serialized

Commands and evidence:

```text
cargo fmt --all -- --check
CARGO_TARGET_DIR=/tmp/loongclaw-target-checkapp cargo check -p loongclaw-app
CARGO_TARGET_DIR=/tmp/loongclaw-target-checkdaemon cargo test -p loongclaw-daemon chat_tui_missing_config_can_finish_fullscreen_onboarding_and_enter_chat -- --test-threads=1
CARGO_TARGET_DIR=/tmp/loongclaw-target-checkdaemon cargo test -p loongclaw-daemon tui_pty --test integration -- --test-threads=1
CARGO_TARGET_DIR=/tmp/loongclaw-target-checkdaemon cargo test -p loongclaw-daemon onboard_cli --lib -- --test-threads=1
CARGO_TARGET_DIR=/tmp/loongclaw-target-fullverify cargo clippy --workspace --all-targets --all-features -- -D warnings
CARGO_TARGET_DIR=/tmp/loongclaw-target-fullverify cargo test --workspace --locked
CARGO_TARGET_DIR=/tmp/loongclaw-target-fullverify cargo test --workspace --all-features --locked
scripts/check_architecture_boundaries.sh
scripts/check_dep_graph.sh
diff CLAUDE.md AGENTS.md

cargo fmt --all -- --check
CARGO_TARGET_DIR=/tmp/loongclaw-target-checkapp-full cargo test -p loongclaw-app
CARGO_TARGET_DIR=/tmp/loongclaw-target-checkdaemon cargo test -p loongclaw-daemon tui_pty --test integration -- --test-threads=1
```

Results:
- All commands above passed.
- `task` was not installed locally, so the touched-area architecture and dep-graph gates were run through the repository scripts directly instead of `task check:*`.
- The focused PTY regression coverage now includes missing-config onboarding entry, first-run completion, the handoff into the fullscreen chat shell, and a startup synchronization guard that waits for the first visible fullscreen frame before onboarding-copy assertions.
- The new happy-path PTY test injects `OPENAI_CODEX_OAUTH_TOKEN` into the child process only. Existing daemon-side env-mutating onboarding tests continue to serialize process-global env access through `ScopedEnv`.

## User-visible / Operator-visible Changes

- `loongclaw chat --ui tui` now opens a real transcript-first fullscreen shell instead of a guarded preview fallback.
- When config is missing, the first-run setup stays inside that fullscreen shell and transitions directly into chat after a successful write.
- Tool/runtime visibility, transcript rendering, and the composer now stay within one app-owned TUI surface across the first-run path.
- The fullscreen chat shell now shows clearer transcript state: multiline compose works, transcript scrolling is keyboard-navigable when the composer is empty, and scrolled-away views visibly advertise both `End latest` and `SCROLLED` state.

## Failure Recovery

- Fast rollback or disable path:
  Revert commit `26eab5e1` and follow-up commit `4c316e78` to remove the fullscreen first-run handoff and transcript polish while preserving the earlier branch baseline, or use `loongclaw chat --ui text` / `loongclaw chat` as the compatibility path.
- Observable failure symptoms reviewers should watch for:
  Missing-config TUI launches dropping back to plain onboarding, successful onboarding failing to hand off into chat, fullscreen startup getting stuck on a blank alternate screen, or transcript navigation keys interfering with multiline compose.

## Reviewer Focus

- `crates/app/src/chat/tui/boot.rs`, `crates/app/src/chat/tui/mod.rs`, and `crates/app/src/chat/tui/shell.rs` for lazy boot integration, composer/input handling, and boot-to-chat transition behavior.
- `crates/app/src/chat/tui/history.rs`, `crates/app/src/chat/tui/status_bar.rs`, and `crates/app/src/chat/tui/render.rs` for transcript affordances, live/scrolled state signaling, and shortcut discoverability.
- `crates/daemon/src/tui_cli.rs`, `crates/daemon/src/lib.rs`, and `crates/daemon/src/onboard_cli/fullscreen.rs` for missing-config routing and the first-run fullscreen onboarding state machine.
- `crates/daemon/tests/integration/tui_pty.rs` for PTY coverage of fullscreen onboarding entry, review/write flow, chat handoff, and startup synchronization.

### CODE REVIEW

File: crates/app/src/chat.rs [hunk 1]
```diff
@@ -21,8 +21,27 @@ use crate::acp::{
 use crate::context::{DEFAULT_TOKEN_TTL_S, bootstrap_kernel_context_with_config};
 
 mod cli_input;
+mod live_surface;
+mod text_surface;
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/app/src/chat.rs:25]: Fixed in 703e69ac: `mod tui` is now gated behind `channel-cli`, the `CliChatUiMode::Tui` branch is cfg-guarded, and `--ui tui` degrades back to text when the feature is absent. The CI `rust-check-browser-without-webfetch` feature mix passes locally.

File: crates/app/src/chat/tui/execution_drawer.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        return Some(DrawerPayload::new(
+            DrawerPayloadKind::ErrorDetail,
+            "Error detail",
+            snapshot.tool_activity_lines.clone(),
+        ));
+    }
+
+    None
+}
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/app/src/chat/tui/execution_drawer.rs]: Fixed in 703e69ac: drawer approval/error detection now filters out raw `args:` lines before classification.

File: crates/app/src/chat/tui/app_shell.rs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        KeyCode::Backspace => Some(UiEvent::Backspace),
+        KeyCode::Tab => Some(UiEvent::FocusDrawer),
+        KeyCode::BackTab => Some(UiEvent::FocusComposer),
+        KeyCode::Char(ch) if !event.modifiers.intersects(KeyModifiers::CONTROL) => {
+            Some(UiEvent::ComposerInput(ch))
+        }
+        KeyCode::Enter
+        | KeyCode::Left
+        | KeyCode::Right
+        | KeyCode::Up
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/app/src/chat/tui/app_shell.rs]: Scoped in 703e69ac: `--ui tui` no longer enters the placeholder shell. It now warns and falls back to the text UI until the runtime-backed shell is wired, so this non-submitting scaffold is not user-reachable in this PR.

File: crates/app/src/chat/live_surface.rs [hunk 4]
```diff
@@ -0,0 +1,898 @@
...
+fn should_render_cli_chat_live_phase(phase: ConversationTurnPhase) -> bool {
+    match phase {
+        ConversationTurnPhase::Preparing
+        | ConversationTurnPhase::RequestingProvider
+        | ConversationTurnPhase::RunningTools
+        | ConversationTurnPhase::RequestingFollowupProvider
+        | ConversationTurnPhase::FinalizingReply
+        | ConversationTurnPhase::Failed => true,
+        ConversationTurnPhase::ContextReady | ConversationTurnPhase::Completed => false,
+    }
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/app/src/chat/live_surface.rs:244]: Fixed in 703e69ac: `ConversationTurnPhase::Completed` now renders through `should_render_cli_chat_live_phase`, so successful turns emit a final live-surface snapshot.

File: crates/app/src/chat/tui/mod.rs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+pub(super) async fn run_tui_chat(
+    runtime: &super::CliTurnRuntime,
+    _options: &super::CliChatOptions,
+) -> CliResult<CliTuiLaunchResult> {
+    let policy =
+        terminal::resolve_terminal_policy(terminal::TerminalSupportSnapshot::capture_current());
+    match policy.launch {
+        terminal::TerminalLaunch::Tui => {
+            app_shell::run_placeholder_shell(runtime, policy.use_plain_palette).await?;
+            Ok(CliTuiLaunchResult::Handled)
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/app/src/chat/tui/mod.rs]: Fixed in 703e69ac: supported terminals no longer return `Handled` from the placeholder shell. `--ui tui` now emits an explicit fallback to the working text UI until the runtime-backed shell is wired.

File: crates/app/src/chat/tui/app_shell.rs [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+                terminal.draw(&state, palette)?;
+            }
+            Ok(Event::Resize(_, _)) => terminal.draw(&state, palette)?,
+            Ok(_) => {}
+            Err(error) => return Err(format!("failed to read TUI input event: {error}")),
+        }
+    }
+
+    Ok(())
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/app/src/chat/tui/app_shell.rs]: Scoped in 703e69ac: this PR no longer routes `--ui tui` into the placeholder shell. Supported terminals now fall back to the working text UI until the runtime-backed shell exists, and the PR/README scope has been updated to match that.

File: crates/app/src/chat/tui/execution_band.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        .iter()
+        .rev()
+        .find(|line| !line.starts_with("args:") && !line.starts_with("[running]"))
+        .cloned();
+
+    ExecutionBandSummary {
+        running_count,
+        pending_approval_count: 0,
+        latest_result,
+        background_count: 0,
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/app/src/chat/tui/execution_band.rs]: Fixed in 703e69ac: the execution band now counts only structured approval status lines and `latest_result` only considers bracketed non-running status rows, so raw args/hint lines no longer leak into the quiet summary.

File: crates/app/src/chat/tui/app_shell.rs [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+                    continue;
+                };
+
+                if reduce(&mut state, ui_event) {
+                    break;
+                }
+
+                terminal.draw(&state, palette)?;
+            }
+            Ok(Event::Resize(_, _)) => terminal.draw(&state, palette)?,
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/app/src/chat/tui/app_shell.rs]: Not wiring this placeholder-specific focus behavior in this PR. In 703e69ac the CLI stops entering `app_shell` for `--ui tui` and falls back to text until runtime integration exists, so this scaffold is no longer user-reachable here.

**Comment by gh-xj** [comment]
LoongClaw AI scientist slot `curie` has claimed this case and is running review execution.

**Comment by gh-xj** [comment]
## LoongClaw QA Review — PR #691

**Reviewed commit**: `703e69ac`
**Risk**: high
**Slot**: `curie`

### Findings

- **MEDIUM**: Dual crossterm versions (0.28.1 + 0.29.0) in lockfile — ratatui 0.29 pins crossterm 0.28 while workspace pins 0.29. Both register SIGWINCH handlers. Currently test-only risk (`app_shell.rs` is `#[cfg(test)]`-gated), but will become production risk when TUI promotes. Consider aligning when ratatui 0.30 ships.
- **MEDIUM**: Signal detection false positives — `line_signals_error`/`line_signals_approval` in `execution_drawer.rs` use broad substring matching. Mitigated by `#[cfg(test)]`-only compilation. Tighten to prefix-based matching before TUI promotion.
- **MEDIUM**: `build_cli_chat_options` public API signature changed (added `ui: CliChatUiModeArg` first param). Only known consumer (daemon) updated. Semver break if library API is consumed externally.
- **LOW**: `pub(crate)` items inside `pub(super)` test-only modules — unnecessary coupling surface
- **LOW**: `#![allow(dead_code)]` blanket suppression — appropriate for groundwork, narrow before promotion
- **LOW**: Stale `RUSTSEC-2025-0057` advisory ignore in `deny.toml` — cleanup recommended
- **LOW**: `CliChatUiMode` is `pub` (fully public) — `pub(crate)` would suffice

### Coverage

- Rust-specific review: applied (module visibility, feature gating, RAII, Elm-style state, dep management)
- Harness review: applied (fmt, clippy, deny, tests, architecture checks, dep graph — all pass)
- QA probes: `toolcall-final` + `tool-loop-safety` — both pass
- Adversarial challenge: applied — identified dual crossterm versions and pub API break (missed by primary review)

### External Reviewer Findings

All Copilot and CodeRabbit Critical/Major findings were addressed in the reviewed commit (703e69ac). No net-new valid findings remain.

### Open Questions

- Will dual crossterm versions be aligned before TUI promotion? (ratatui 0.30 targets crossterm 0.29)
- Is `build_cli_chat_options` consumed outside the workspace?
- What is the TUI promotion plan from `#[cfg(test)]` to production?

### Verdict

PR is sound for its declared scope (groundwork/fallback only). No production-impacting bugs found. The key follow-up risk is the dual crossterm versions which should be addressed before TUI code promotes from `#[cfg(test)]` to production. The chat.rs extraction is structurally clean, feature gating is correct, and all verification checks pass.

---
*ai-scientist `curie` · case CASE-20260328-pr691-tui-chat-groundwork*

**Comment by gh-xj** [comment]
## Status Bar Token Fix

Pushed `e6e1bb4a` — fixes the status bar stuck at "0 tokens (0%)" permanently.

**Root cause:** Three breaks in the data pipeline:
- `observer.rs` hardcoded `output_tokens: 0`
- `ConversationTurnPhaseEvent` lacked actual token fields
- Provider response usage was never extracted from `raw_meta`

**Fix:** Added `ProviderUsage` struct, extract usage from provider response body (OpenAI/Anthropic/Bedrock), thread `actual_input_tokens`/`actual_output_tokens` through the event pipeline, observer prefers actual values with fallback to estimates.

**Verified:** 2488 tests pass (2 pre-existing failures unrelated), clippy clean, manual QA confirms status bar updates to real token counts after provider response.

**Remaining:** `context_length` is never set from config, so percentage stays at 0%. Tracked separately.

**Comment by gh-xj** [comment]
## TUI Polish Batch (this session)

### Commits added:

| Commit | Description |
|--------|-------------|
| `e6e1bb4a` | **Status bar tokens**: wire provider usage through event pipeline |
| `a998463e` | **Light theme**: redesign palette for proper contrast |
| `5b151ca2` | **Context length**: infer from model name + **Staged queue**: depth-1 message queue for safe turn submission |
| `74552736` | **Markdown rendering**: custom pulldown-cmark renderer (headers, bold, lists, code, blockquotes) |
| `b951eebd` | Enable pulldown-cmark default features |

### Issues resolved:
- #712 — context_length / percentage display
- #713 — turn submission race → staged message queue
- #714 — light theme contrast
- #715 — markdown rendering in assistant responses

### Verification:
- 2532 tests pass (1 pre-existing failure in history divider test, unrelated)
- All pre-commit hooks (fmt + clippy) green on every commit
- Manual QA: status bar shows real token counts + percentage, markdown renders without raw syntax, light theme readable, staged queue works

**Comment by chumyin** [comment]
Superseded by #866.

This replacement PR carries the current fullscreen onboarding + chat TUI delivery line on branch `ref/onboard-chat-to-full-tui`, including the `#690` hero-first onboarding integration, transcript review/tool-visibility refinements, explicit onboarding-to-chat session continuity, and the final retry-test stabilization needed to get the full validation set green.

Keeping #691 open for traceability until the replacement PR lands.

**Comment by Mike-7777777** [comment]
FYI: PR #1054 (pending review) changes the default home directory from \`~/.loongclaw\` to \`~/.loong\` and adds \`HOME_DIR_NAME\` / \`LEGACY_HOME_DIR_NAME\` constants in \`config::shared\`. If this TUI onboarding PR references the home directory path (e.g., displaying where config will be saved, or creating files under the home), it may need to use these constants or account for the new default when rebasing onto \`dev\` after #1054 merges.

**Comment by chumyin** [comment]
closing this as outdated. we are redesigning the onboard flow and the tui ui/ux, so this is no longer the direction we want to keep open. we can reopen the work in a fresh pr once the new design is settled.
